### PR TITLE
[RDY] Refactor overmap specials placing and fix some bugs (mapgen)

### LINKED ISSUE #17475: Mapgen: Some buildings don't spawn by a road
Most of the time, it's mansions, but I've also seen it with evac shelters, apartment towers, FEMA camps, military outposts (intended?), schools, and toxic waste dumps (intended?). I have screenshots if anyone is interested.

### LINKED ISSUE #17845: by_hiway flag is broken
I gather that hiways are not a priority, but adding anything larger than a 1x1 building t the spawn location "by_hiway" results in it spawning cradled or surrounded by river.

perhaps we can find a use for freeways?  if 3d is still in earnest devlopment a raised freeway is entirely possible for cities and it could work for intercity routes?  itd be nice  to know which roads conected to other cities instead of a farm 35 tiles out.

### LINKED ISSUE #18726: Roads overwriting building segments
It seems that in some cases the roads that are supposed to connect building and towns may overwrite parts of the building.

Readily evident with the apple orchard, football field, and large power substation; but can also happen with the apartment building

### PR DESCRIPTION
**Pull request by codemime**
Fixes #17475.
Fixes #17845.
Fixes #18726.

The #18726 bug is in `overmap::place_special()` and it's quite old. The function assumes that the road should come out of the "top" of the map specials, and if it's not the case, it places the road without any rotation and regardless of  the `"allow_road"` flag.

**Before (note the overwritten corners of football fields and absence of schools):**
before2

**After:**
after_2

Be aware that these examples are purely synthetic: for debugging and merge-testing purposes.
To reproduce them replace [here](https://github.com/CleverRaven/Cataclysm-DDA/commit/2af2d4bea014168baf97c8d08d1832289c571de9#diff-90df0fb487770203c1f3f21de7b1b6bdR3565):

``` c++
if( ( num_placed[&special] < special.max_occurrences || special.max_occurrences <= 0 ) &&
    allow_special( special, p, rotation ) ) {
```

with something like:

``` c++
if( ( special.id == "..." || ... ) && allow_special( special, p, rotation ) ) {
```

### Implementation details

- Naive assumptions that specials are always 1x1 and roads always connect to the northern edge were replaced with more robust algorithm
- No more "forced roads" due to more accurate checks
- More generic behaviour. Any "road" can be connected now (e.g. sewer or something)
- Validity checks were added to overmap specials
- Specials and connections now use native ids
- Unused locations `"always"` and `"never"` were removed
- Location `"by_hiway"` was renamed to `"by_road"`
- Various code refinements

The new code is also [presumably, I didn't compare] somewhat faster due to some optimizations and caching.

### CODE REVIEW

**Comment by John-Candlebury** [comment]
Thanks for the fix!

**Comment by codemime** [comment]
You're welcome. I'll finish it soon.

**Comment by codemime** [comment]
Ready.

File: src/overmap.cpp [hunk 1]
```diff
@@ -75,6 +75,21 @@ oter_iid ot_null,
...
+
+//@todo Jsonize this map.
+static const std::map<std::string, overmap_location_restictions> overmap_locations = {
+    { "by_road",    { {},                   { "river", "road" } } },
+    { "field",      { { "field" },          {}                  } },
+    { "forest",     { { "forest" },         {}                  } },
+    { "land",       { {},                   { "river", "road" } } },
+    { "water",      { { "river" },          {}                  } },
+    { "wilderness", { { "forest", "field" },{}                  } }
+};
```
Comment (Coolthulhu) [inline src/overmap.cpp]: Comment would be useful - it's not obvious from context alone.

File: src/overmap.cpp [hunk 2]
```diff
@@ -121,6 +136,11 @@ city::city( int const X, int const Y, int const S)
 {
 }
 
+int city::get_distance_from( const tripoint &p ) const
+{
+    return std::max( int( trig_dist( p, { x, y, 0 } ) ) - s, 0 );
```
Comment (Coolthulhu) [inline src/overmap.cpp]: Why the `max`? Can it ever be `<0`?

File: src/overmap.cpp [hunk 3]
```diff
@@ -938,6 +963,81 @@ void apply_region_overlay(JsonObject &jo, regional_settings &region)
 
 }
 
+const overmap_special_terrain &overmap_special::get_terrain_at( const tripoint &p ) const
+{
+    for( const auto &elem : terrains ) {
+        if( elem.p == p ) {
+            return elem;
+        }
+    }
```
Comment (Coolthulhu) [inline src/overmap.cpp]: Could be cleaner with a `std::find` or `std::find_if`.

File: src/overmap.cpp [hunk 4]
```diff
@@ -938,6 +963,81 @@ void apply_region_overlay(JsonObject &jo, regional_settings &region)
...
+            points.insert( elem.p );
+        }
+
+        if( !elem.connect ) {
+            continue;
+        }
+
+        bool found = false;
+
+        for( const auto &n : { NORTH, SOUTH, EAST, WEST } ) {
```
Comment (Coolthulhu) [inline src/overmap.cpp]: Will `auto` with an initializer list be unambiguous here? I can easily see some compiler complaining about not being able to figure out what is it.

File: src/overmap.cpp [hunk 5]
```diff
@@ -3909,108 +3994,83 @@ inline tripoint rotate_tripoint(tripoint p, int rotations)
...
-                break;
-            case 2:
-                testpoint = tripoint(testpoint.x, testpoint.y + 1, testpoint.z);
-                break;
-            case 3:
-                testpoint = tripoint(testpoint.x - 1, testpoint.y, testpoint.z);
-                break;
-            default:
+    rotations.reserve( num_rotations );
+    // Try to find the best rotation
```
Comment (Coolthulhu) [inline src/overmap.cpp]: Comment would be useful. One line about what is the score.

File: src/overmap.cpp (hunk 4 above)
Comment (codemime) [inline src/overmap.cpp]: I had inlined it [at some point](https://github.com/CleverRaven/Cataclysm-DDA/pull/18982/commits/1379c92210ee4184a140f12c25fcca2ba1c8cbac#diff-90df0fb487770203c1f3f21de7b1b6bdL966), but didn't see any ill effects. I wonder if there's a modern compiler that would stumble here.

File: src/overmap.cpp (hunk 1 above)
Comment (codemime) [inline src/overmap.cpp]: Added.

**Comment by codemime** [comment]
All set.

**Comment by mugling** [comment]
I keep keep getting notifications for this the past few days - are you rebasing? It's difficult to determine if your actually done?

**Comment by codemime** [comment]
I was squashing changes to the last commit. They didn't deserve a separate one.

**Comment by Coolthulhu** [comment]
Megastore road generation is broken - roads by megastores end in cul de sacs, while ones generated using old mapgen connect nicely.

**Comment by codemime** [comment]
> while ones generated using old mapgen connect nicely.

I'll test it again, but are you sure that it's caused by the PR? Considering the fact that the issue exists and it's still unfixed (I've checked):  #17638.

**Comment by codemime** [comment]
This PR is the first in a chain of further mapgen PRs. I'm overwriting the placing code for the specials. It becomes nicer and faster. I'm gonna fix the #17638 too.

**Comment by Coolthulhu** [comment]
Road requirements aren't enforced properly. I am getting gas stations (the model with 2 garages) spawning by rivers and swamps.

**Comment by Coolthulhu** [comment]
I can't get the small evac center (the default survivor spawn) to generate. The world is all default, nothing changed.

**Comment by codemime** [comment]
> I can't get the small evac center (the default survivor spawn) to generate. The world is all default, nothing changed.

Reproducible. I'm repairing that.

**Comment by codemime** [comment]
In the master branch (prior to this PR) [the gas station](https://github.com/CleverRaven/Cataclysm-DDA/pull/18982/files#diff-159a08a42909fbde9aae992dd5a86751L1509) wasn't marked with `"by_hiway"` flag and could spawn freely. Should I fix that in the following PR?

**Comment by mutability** [comment]
The by_hiway removal seems to have missed updating the "more locations" and "z-level buildings" mods, they generate a bunch of errors on load now.

**Comment by codemime** [comment]
Damn, how could I forget about the mods. Thanks for the notification.

**Pull request merged**